### PR TITLE
[Object Detection] Add DETR Model

### PR DESCRIPTION
**Pull request by DavidLandup0**
Draft PR for observability. Body will be updated as progress is made.

### DETR Elements

Will include:
- DETR layers
    - Positional embeddings
    - Frozen BatchNormalization
    - DETRTransformerEncoder
    - DETRTransformerDecoder
- DetrBackbone (Modified ResNet)
- Object detection preprocessor
- DETRObjectDetector (Tying the above into a class)
- Weight conversion script from https://huggingface.co/facebook/detr-resnet-50

... and relevant workflows, collab notebooks, etc. for sanity checking.

### CODE REVIEW

**Comment by DavidLandup0** [comment]
Getting there :)
Lots of refactoring to do, but the core modules are there and work nicely with each other.

image

**Comment by sachinprasadhs** [comment]
@DavidLandup0 , Any update on this PR?

**Comment by DavidLandup0** [comment]
> @DavidLandup0 , Any update on this PR?

Picking it up now that a couple of other things are out the way - should have a more concrete update soon :)
The architecture ~matched last time I worked on this, so not much should be left.

File: keras_hub/src/utils/transformers/preset_loader.py [hunk 1]
```diff
@@ -109,5 +115,10 @@ def load_tokenizer(self, cls, config_name="tokenizer.json", **kwargs):
         return self.converter.convert_tokenizer(cls, self.preset, **kwargs)
 
     def load_image_converter(self, cls, **kwargs):
+        # Check if the converter has an image converter conversion function
```
Comment (DavidLandup0) [inline keras_hub/src/utils/transformers/preset_loader.py:118]: Adding this for backwards compatability - looks like we just didn't have a way to return the image converter so far, unless I missed it from another model.

This way, any `convert_x.py` that has `convert_image_converter()` method can return that image converter too.

File: keras_hub/src/models/detr/detr_backbone.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    rich features for object detection. The backbone:
+    1. Extracts features using ResNet
+    2. Projects to hidden dimension via 1x1 convolution
+    3. Adds 2D positional embeddings
+    4. Processes through transformer encoder
+
+    Args:
+        image_encoder: keras.Model. CNN backbone, typically
+            ResNetBackbone.from_preset("resnet_50_imagenet")
+        hidden_dim: int. Model dimension (d_model), default 256
```
Comment (sachinprasadhs) [inline keras_hub/src/models/detr/detr_backbone.py:27]: Follow strcture for all defaults --> . Defaults to `256`.

File: keras_hub/src/models/detr/detr_backbone.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        image_encoder,
+        hidden_dim=256,
+        num_encoder_layers=6,
+        num_heads=8,
+        intermediate_size=2048,
+        dropout=0.1,
+        activation="relu",
+        image_shape=(None, None, 3),
+        **kwargs,
+    ):
```
Comment (sachinprasadhs) [inline keras_hub/src/models/detr/detr_backbone.py:63]: Add `dtype` and `data_format` argument.

File: keras_hub/src/models/detr/detr_backbone.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        num_encoder_layers: int. Number of encoder layers, default 6
+        num_heads: int. Number of attention heads, default 8
+        intermediate_size: int. FFN intermediate size, default 2048
+        dropout: float. Dropout rate, default 0.1
+        activation: str. Activation function, default "relu"
+        image_shape: tuple. Input image shape, default (None, None, 3)
+
+    Example:
+    ```python
+    # Create backbone
```
Comment (sachinprasadhs) [inline keras_hub/src/models/detr/detr_backbone.py:37]: May be chnage it as Create DETR backbone or add comment for the flow like Image encoder, DETR backbone, Extract features.

File: keras_hub/src/models/detr/detr_backbone.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    """
+
+    def __init__(
+        self,
+        image_encoder,
+        hidden_dim=256,
+        num_encoder_layers=6,
+        num_heads=8,
+        intermediate_size=2048,
+        dropout=0.1,
```
Comment (sachinprasadhs) [inline keras_hub/src/models/detr/detr_backbone.py:59]: If these are varying from one checkpoint config to another, let's not make this as default values.

File: keras_hub/src/models/detr/detr_backbone_test.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+            "num_encoder_layers": 2,
+            "num_heads": 8,
+            "intermediate_size": 512,
+            "dropout": 0.1,
+            "activation": "relu",
+            "image_shape": (None, None, 3),
+        }
+
+        self.input_size = 256
+        self.input_data = ops.ones((2, self.input_size, self.input_size, 3))
```
Comment (sachinprasadhs) [inline keras_hub/src/models/detr/detr_backbone_test.py:34]: Test it with smallest possible values, so that we don't put extra load on GPUs.

File: keras_hub/src/models/detr/detr_backbone_test.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+            init_kwargs=self.init_kwargs,
+            input_data=self.input_data,
+            expected_output_shape={
+                "encoded_features": (2, 256, 256),
+                "pos_embed": (2, 256, 256),
+                "mask": (2, 256),
+            },
+            run_mixed_precision_check=False,
+            run_quantization_check=False,
+            run_data_format_check=False,
```
Comment (sachinprasadhs) [inline keras_hub/src/models/detr/detr_backbone_test.py:48]: Enable these tests, since resnet is having these enabled, you need to handle only in your model implementation.

File: keras_hub/src/models/detr/detr_image_converter.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+                -0.456 / 0.224,  # G channel
+                -0.406 / 0.225,  # B channel
+            ]
+
+        super().__init__(
+            image_size=image_size,
+            scale=scale,
+            offset=offset,
+            **kwargs,
+        )
```
Comment (sachinprasadhs) [inline keras_hub/src/models/detr/detr_image_converter.py:38]: We don't need this here.. it will be handled here https://github.com/keras-team/keras-hub/blob/master/keras_hub/src/layers/preprocessing/image_converter.py#L113

File: keras_hub/src/models/detr/detr_layers.py [hunk 9]
```diff
@@ -0,0 +1,1068 @@
...
+            initializer="zeros",
+            trainable=False,
+            name="running_mean",
+        )
+        self.running_var = self.add_weight(
+            shape=(self.num_features,),
+            initializer="ones",
+            trainable=False,
+            name="running_var",
+        )
```
Comment (sachinprasadhs) [inline keras_hub/src/models/detr/detr_layers.py:43]: This is PyTorch naming convention.. we use gamma for weight, beta for bias, moving_mean for running_mean and moving_variance for running_var.

File: keras_hub/src/models/detr/detr_layers.py [hunk 10]
```diff
@@ -0,0 +1,1068 @@
...
+        pos = ops.transpose(pos, [0, 3, 1, 2])
+        return pos
+
+    def compute_output_shape(self, input_shape):
+        """Compute output shape: (batch, embedding_dim*2, height, width)."""
+        batch_size, height, width = input_shape
+        return (batch_size, self.embedding_dim * 2, height, width)
+
+
+def position_embedding_sine(
```
Comment (sachinprasadhs) [inline keras_hub/src/models/detr/detr_layers.py:118]: Where is this getting used?

**Review by sachinprasadhs — COMMENTED** [review]
Thanks. Took a look into few files, made some comments.
